eth_trie/hexary.nim

```diff
@@ -119,16 +122,219 @@ proc getAux(db: DB, nodeRlp: Rlp, path: NibblesRange): BytesRange =
+    if not lastElem.isEmpty:
+      leaves.add lastElem.toBytes
+  else:
+    raise newException(CorruptedTrieError,
+                       "HexaryTrie node with an unexpected number of children")
+
+proc getValues*(self: HexaryTrie): seq[BytesRange] =
+  result = @[]
+  var nodeRlp = rlpFromBytes keyToLocalBytes(self.db, self.root)
+  self.db.getValuesAux(nodeRlp, result)
+
+proc getKeysAux(db: DB, nodeRlp: Rlp, keys: var seq[ByteRange], path: NibblesRange) =
+  if not nodeRlp.hasData or nodeRlp.isEmpty:
+    return
+
+  case nodeRlp.listLen
+  of 2:
+    let
+      (isLeaf, k) = nodeRlp.extensionNodeKey
+      key = path & k
+    if isLeaf:
+      assert(key.len mod 2 == 0)
+      keys.add key.getBytes
+    else:
+      let
+        value = nodeRlp.listElem(1)
+        nextLookup = value.getLookup
+      db.getKeysAux(nextLookup, keys, key)
+  of 17:
+    var key = path.cloneAndReserveNibble()
+    for i in 0 ..< 16:
+      var branch = nodeRlp.listElem(i)
+      if not branch.isEmpty:
+        let nextLookup = branch.getLookup
+        key.replaceLastNibble(i.byte)
+        db.getKeysAux(nextLookup, keys, key)
+
+    var lastElem = nodeRlp.listElem(16)
+    if not lastElem.isEmpty:
+      assert(path.len mod 2 == 0)
+      keys.add path.getBytes
+  else:
+    raise newException(CorruptedTrieError,
+                       "HexaryTrie node with an unexpected number of children")
+
+proc getKeys*(self: HexaryTrie): seq[BytesRange] =
+  result = @[]
+  var nodeRlp = rlpFromBytes keyToLocalBytes(self.db, self.root)
+  var path = newRange[byte](0)
+  self.db.getKeysAux(nodeRlp, result, initNibbleRange(path))
+
+proc getKeysAux(db: DB, stack: var seq[tuple[nodeRlp: Rlp, path: NibblesRange]]): BytesRange =
+  while stack.len > 0:
+    let (nodeRlp, path) = stack.pop()
+    if not nodeRlp.hasData or nodeRlp.isEmpty:
+      continue
+
+    case nodeRlp.listLen
+    of 2:
+      let
+        (isLeaf, k) = nodeRlp.extensionNodeKey
+        key = path & k
+
+      if isLeaf:
+        assert(key.len mod 2 == 0)
+        return key.getBytes
+      else:
+        let
+          value = nodeRlp.listElem(1)
+          nextLookup = value.getLookup
+        stack.add((nextLookup, key))
+    of 17:
+      for i in 0 ..< 16:
+        var branch = nodeRlp.listElem(i)
+        if not branch.isEmpty:
+          let nextLookup = branch.getLookup
+          var key = path.cloneAndReserveNibble()
+          key.replaceLastNibble(i.byte)
+          stack.add((nextLookup, key))
+
+      var lastElem = nodeRlp.listElem(16)
+      if not lastElem.isEmpty:
+        assert(path.len mod 2 == 0)
+        return path.getBytes
+    else:
+      raise newException(CorruptedTrieError,
+                        "HexaryTrie node with an unexpected number of children")
+
+iterator keys*(self: HexaryTrie): BytesRange =
+  var
+    nodeRlp = rlpFromBytes keyToLocalBytes(self.db, self.root)
+    path = newRange[byte](0)
+    stack = @[(nodeRlp, initNibbleRange(path))]
```

Comment:
On first glance, I think the maximum size of this stack is bounded to `16 (max branches per level) * 64 (max levels)`, so we can consider using stack memory for it, but even if we don't do this, it would be nice to pre-allocate a `seq` with a bit larger capacity here.

Reply:
The more complicated stack-based version for the iterator is not slower, right? Why don't we remove the recursive version?